packages/std/src/math/int512.rs

```diff
@@ -0,0 +1,1350 @@
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        // The inner type doesn't work as expected with padding, so we
+        // work around that. Remove this code when the upstream padding is fixed.
+        let unpadded = self.0.to_string();
+        let numeric = unpadded.strip_prefix('-').unwrap_or(&unpadded);
+
+        f.pad_integral(self >= &Self::zero(), "", numeric)
+    }
+}
+
+impl Add<Int512> for Int512 {
+    type Output = Self;
+
+    fn add(self, rhs: Self) -> Self {
+        Int512(self.0.checked_add(rhs.0).unwrap())
+    }
+}
+
+impl<'a> Add<&'a Int512> for Int512 {
+    type Output = Self;
+
+    fn add(self, rhs: &'a Int512) -> Self {
+        Int512(self.0.checked_add(rhs.0).unwrap())
+    }
+}
+
+impl Sub<Int512> for Int512 {
+    type Output = Self;
+
+    fn sub(self, rhs: Self) -> Self {
+        Int512(self.0.checked_sub(rhs.0).unwrap())
+    }
+}
+forward_ref_binop!(impl Sub, sub for Int512, Int512);
+
+impl SubAssign<Int512> for Int512 {
+    fn sub_assign(&mut self, rhs: Int512) {
+        self.0 = self.0.checked_sub(rhs.0).unwrap();
+    }
+}
+forward_ref_op_assign!(impl SubAssign, sub_assign for Int512, Int512);
+
+impl Div<Int512> for Int512 {
+    type Output = Self;
+
+    fn div(self, rhs: Self) -> Self::Output {
+        Self(self.0.checked_div(rhs.0).unwrap())
+    }
+}
+
+impl<'a> Div<&'a Int512> for Int512 {
+    type Output = Self;
+
+    fn div(self, rhs: &'a Int512) -> Self::Output {
+        Self(self.0.checked_div(rhs.0).unwrap())
+    }
+}
+
+impl Rem for Int512 {
+    type Output = Self;
+
+    /// # Panics
+    ///
+    /// This operation will panic if `rhs` is zero.
+    #[inline]
+    fn rem(self, rhs: Self) -> Self {
+        Self(self.0.rem(rhs.0))
+    }
+}
+forward_ref_binop!(impl Rem, rem for Int512, Int512);
+
+impl Not for Int512 {
+    type Output = Self;
+
+    fn not(self) -> Self::Output {
+        Self(!self.0)
+    }
+}
+
+impl Neg for Int512 {
+    type Output = Self;
+
+    fn neg(self) -> Self::Output {
+        Self(-self.0)
+    }
+}
+
+impl RemAssign<Int512> for Int512 {
+    fn rem_assign(&mut self, rhs: Int512) {
+        *self = *self % rhs;
+    }
+}
+forward_ref_op_assign!(impl RemAssign, rem_assign for Int512, Int512);
+
+impl Mul<Int512> for Int512 {
+    type Output = Self;
+
+    fn mul(self, rhs: Self) -> Self::Output {
+        Self(self.0.checked_mul(rhs.0).unwrap())
+    }
+}
+forward_ref_binop!(impl Mul, mul for Int512, Int512);
+
+impl MulAssign<Int512> for Int512 {
+    fn mul_assign(&mut self, rhs: Self) {
+        self.0 = self.0.checked_mul(rhs.0).unwrap();
+    }
+}
+forward_ref_op_assign!(impl MulAssign, mul_assign for Int512, Int512);
+
+impl Shr<u32> for Int512 {
+    type Output = Self;
+
+    fn shr(self, rhs: u32) -> Self::Output {
+        self.checked_shr(rhs).unwrap_or_else(|_| {
+            panic!(
+                "right shift error: {} is larger or equal than the number of bits in Int512",
+                rhs,
+            )
```

Comment:
Is this panic how standard types behave? Why not just move all bits out of the type and be happy?

Reply:
Yes, it is: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=00f8f81172a40ce26b8d5b7e9ae9d96b

Reply:
Okay now I am consused. Why do they call this overflow ("attempt to shift left with overflow")? Isn't that an invalid input panic instead of an overflow panic?

I can do `println!("{}", u32::MAX << 1);` which does not panic so overflowing cannot be the point here.

Reply:
Yes, it's quite confusing. And apparently there is no version of `*_shl` that actually returns 0 (which is what I would expect for unsigned bit shift). There is `overflowing_shl` and `wrapping_shl`, but both just return the input if you shift by the number of bits.
There is also `unchecked_shl` which is unsafe and has UB if the shift amount is >= bit width (apparently c has that behaviour)

Apparently rust just defines it to be an overflow: https://doc.rust-lang.org/reference/expressions/operator-expr.html#overflow 🤷

Reply:
Thank you for the research and confirmation